patina_dxe_core/src/tpl_mutex.rs

```diff
@@ -0,0 +1,239 @@
+//!
+//! This module raises and lowers the UEFI TPL level as the primary means of
+//! guarding the mutex critical section.
+//!
+//! This mutex guarantees that the critical section protected by the guard
+//! cannot be interrupted by code running at TPL equal to or lower than the
+//! lock's TPL level. At TPL_HIGH_LEVEL, interrupts are disabled, so that
+//! means that the critical section cannot be interrupted by anything.
+//!
+//! ## License
+//!
+//! Copyright (c) Microsoft Corporation.
+//!
+//! SPDX-License-Identifier: Apache-2.0
+use core::{
+    cell::UnsafeCell,
+    fmt,
+    ops::{Deref, DerefMut},
+};
+
+use r_efi::efi;
+
+use crate::events::{raise_tpl, restore_tpl};
+
+/// Used to guard data with a locked MUTEX and TPL level.
+pub struct TplMutex<T: ?Sized> {
+    tpl_lock_level: efi::Tpl,
+    lock: UnsafeCell<bool>,
+    name: &'static str,
+    data: UnsafeCell<T>,
+}
+/// Wrapper for guarded data, which can be accessed by Deref or DerefMut on this object.
+pub struct TplGuard<'a, T: ?Sized + 'a> {
+    release_tpl: efi::Tpl,
+    mutex: &'a TplMutex<T>,
+}
+
+unsafe impl<T: ?Sized + Send> Sync for TplMutex<T> {}
+unsafe impl<T: ?Sized + Send> Send for TplMutex<T> {}
+
+unsafe impl<T: ?Sized + Sync> Sync for TplGuard<'_, T> {}
+unsafe impl<T: ?Sized + Send> Send for TplGuard<'_, T> {}
+
+impl<T> TplMutex<T> {
+    /// Instantiates a new TplMutex with the given TPL level, data object, and name string.
+    pub const fn new(tpl_lock_level: efi::Tpl, data: T, name: &'static str) -> Self {
+        Self { tpl_lock_level, lock: UnsafeCell::new(false), data: UnsafeCell::new(data), name }
+    }
+}
+
+impl<T: ?Sized> TplMutex<T> {
+    /// Lock the TplMutex and return a TplGuard object used to access the data. This will raise the system TPL level
+    /// to the level specified at TplMutex creation.
+    ///
+    /// # Panics
+    ///
+    /// Lock re-entrance is not supported; attempt to re-lock something already locked will panic.
+    ///
+    /// Attempting to acquire the lock while running at a TPL level higher than the lock's TPL level will panic due to
+    /// TPL inversion.
+    pub fn lock(&self) -> TplGuard<'_, T> {
+        self.try_lock().unwrap_or_else(|| panic!("Re-entrant locks for {:?} not permitted.", self.name))
+    }
+
+    /// Attempts to lock the TplMutex, and if successful, returns a guard object that can be used to access the data.
+    ///
+    /// # Panics
+    ///
+    /// Attempting to acquire the lock while running at a TPL level higher than the lock's TPL level will panic due to
+    /// TPL inversion.
+    pub fn try_lock(&self) -> Option<TplGuard<'_, T>> {
+        let release_tpl = raise_tpl(self.tpl_lock_level);
+        // Safety: raw pointer on lock is used for volatile semantics; pointer is valid by construction, and exclusive
+        // access is guaranteed by TPL raising.
+        if !unsafe { self.lock.get().read_volatile() } {
+            // Safety: see prior comment.
```

Comment:
I think there is still a very niche double-entry scenario, though let me know if i'm missing something. If we have a TpmMutex at less than HIGH, it's still valid to raise the TPL to the same as the current value. So, if a caller was taking TplMutex.try_lock() on less than HIGH and took an interrupt between lines 76 and 79 and at some point in the interrupt handler tries to acquire the same lock it would succeed and enter the critical section. If it failed to release the lock before returning, then this would allow double entry.

However, I think this is fine if we assume that all locks will be released before the interrupt handler returns, but it might be worth calling out that assumption in the safety comment.

Reply:
I would feel more comfortable with the atomic honestly, if there is no perf difference, but I'm not sure where we landed on that.

Reply:
I put back the atomic. measured performance delta is in the noise - less than 10ms to overall boot time which is in the boot-to-boot variance.